tests: Add criterion for basic benchmarking

Herschel opened the pull request:
Add `criterion` with some basic benchmark SWFs, and refactor some of the test code to support this.

`criterion` seems geared towards smaller benchmarks, and these benchmarks are already a little chunky -- it's possible something like [bencher](https://github.com/bluss/bencher) might be more suitable.

TODO, probably future PRs:
 * Benchmark AVM more directly
    * It'd be nice to benchmark AVM code directly instead of running the entire frame loop & player machinery.
    * Trickier with AVM1 because it's so coupled to the timeline and displaylist.
    * This should be good enough for comparisons for now.
 * `TestBuilder` could become a more general `PlayerBuilder` available in `core`.
 * Get rid of the giant list of tests in `regression_tests.rs` and instead generate tests via directory listing similar to the benchmarks
    * Tests that need a specific config (approx, frame count, etc.) can be specified, otherwise can be omitted.
    * I think this needs to be done via procmacro or build.rs for tests.
 * ~Add some AVM2 and general playback benchmarks.~

Herschel committed:
tests: Refactor test harness

Add `TestBuilder` to allow for configuring the player.

Herschel committed:
benches: Add criterion for benchmarks

adrian17 commented:
> Tests that need a specific config (approx, frame count, etc.)

Could also put test-specific configs as special comments in output.txt. (I'm thinking of something like LLVM's tests.)

But also... are there cases where we want different frame count than the number of frames in the test SWF itself?

adrian17 commented:
Also, let's maybe rename the benches to make it explicit they test AVM1?

Herschel commented:
> Could also put test-specific configs as special comments in output.txt. (I'm thinking of something like LLVM's tests.)

I'm working towards a more general test + SWF compiler [over here](https://github.com/Herschel/rust-as/blob/main/test.rfl), and I figure these settings can be a part of that syntax. But in the meantime doing some YAML front matter in output.txt would work nicely.

Herschel commented:
Currently the results seem to vary quite a lot across runs, so I'm going to try to narrow it down to run some of the AVM code directly.

adrian17 commented:
`large_array_init` feels small enough that it might be beneficial to even wrap the array creation in an ActionScript loop, to minimize startup overhead.

Herschel committed:
benches: Add AVM2 benchmark group

Herschel committed:
benches: Add some AVM2 benchmarks

Herschel committed:
core: Add `PlayerBuilder::with_external_interface`

Remove `Player::add_external_interface` now that these can be
specified at constructor time.

zamazan4ik commented:
Hey, @Herschel! What is the status of this PR?
